[HIBIKE] python protocol changes and runtime integration

karthik-shanmugam opened the pull request:
- change protocol to generalize all devices as param value mappings
    - implement protocol changes in python ONLY
- implement virtual devices for testing
- implement hibike_process.py and hibike_simulator.py for concurrency and runtime integration

karthik-shanmugam committed:
[HIBIKE] add plans for hibike

- add suggestions for protocol changes to readme
- add tasks to todo
- add python software architecture diagram

karthik-shanmugam committed:
[HIBIKE] add hibike atalanta integration spec readme

karthik-shanmugam committed:
[HIBIKE] add hibike_process.py

karthik-shanmugam committed:
[HIBIKE] add support for python3

karthik-shanmugam committed:
[HIBIKE] document new protocol

- redefine all message types in readme
- add skeleton code to hibike_message.py
- update config file to match new protocol

karthik-shanmugam committed:
[HIBIKE] create virtual device skeleton

ivanache committed:
[HIBIKE] Started making the virtual device operational

97hongjun committed:
[HIBIKE] implement some packet constructors in python

karthik-shanmugam committed:
[HIBIKE] add test script to print all devices

ivanache committed:
[HIBIKE] Update virtual device to new protocol

karthik-shanmugam committed:
[HIBIKE] test firmware-python integration

- create python test script to subscribe to all params of devices
- add utility functions to hibike_message.py
- implement a blocking message read function in python

karthik-shanmugam committed:
[HIBIKE] implement hibike process modeled after simulator

karthik-shanmugam committed:
[HIBIKE] implement hibike simulator for integration

97hongjun committed:
[HIBIKE] Add implementation for decoding device write.

ivanache committed:
[HIBIKE] Implement read/write functionality in virtual devices

karthik-shanmugam committed:
[HIBIKE] add script to spawn virtual devices

karthik-shanmugam committed:
[HIBIKE] clean up for merge

karthik-shanmugam committed:
[HIBIKE] clarify python bitwise operations in decoding params

karthik-shanmugam committed:
[HIBIKE] clean up hibike_process.py

karthik-shanmugam committed:
[HIBIKE] clean up from dropping cpp commits

karthik-shanmugam committed:
[HIBIKE] debug virtual device spawn script

karthik-shanmugam committed:
[HIBIKE] use json config in virtual device

nikitakit commented on hibike/ATALANTA_INTEGRATION.md:
```diff
@@ -0,0 +1,83 @@
+
+## StateManager -> Hibike
+
+`["ready"]`
```
ready message is deprecated

nikitakit commented on hibike/ATALANTA_INTEGRATION.md:
```diff
@@ -0,0 +1,83 @@
+
+MetalDetector
+
+- callibrate: bool write only
```
s/callibrate/calibrate/g (replace it throughout this changeset!)

nikitakit commented on hibike/TODO.md:
```diff
@@ -0,0 +1,46 @@
+
+## Protocol Updates (hibike/protocol2.0)
+
+Design, document and implementat the next version of the hibike protocol. See README.md in the hibike/protocol2.0 branch.
```
The comment about the `hibike/protocol2.0` branch is no longer valid.

nikitakit replied:
Also, perhaps this TODO file should be converted to issues on the GitHub issue tracker...

yizow replied:
Oh, and this TODO list is in the shared google doc right?

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -112,31 +114,152 @@ def checksum(data):
+  for param in converted_params:
+     if param >= len(devices[device_id]["params"]):
+        break
+     named_params.append(devices[device_id]["params"][param]["name"])
```
(copied comment from previous PR)

BTW, you can combine these two loops

``` python
res = []
for param_count, param_info in enumerate(devices[device_id]["params"]):
    if params_bitmask & (1 << param_count):
        res.append(param_info["name"])
return res
```

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -112,31 +114,152 @@ def checksum(data):
+      params_and_values - an iterable of param (name, value) tuples
+  """
+  params_and_values = sorted(params_and_values, key=lambda x: paramMap[device_id][x[0]][0])
+  params = [param[0] for param in params_and_values]
```
(copied comment by @yizow from the previous PR)

`params, values = zip(*params_and_values)`

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -112,31 +114,152 @@ def checksum(data):
+      device_id         - a device type id (not uid).
+      params_and_values - an iterable of param (name, value) tuples
+  """
+  params = [param_tuple[0] for param_tuple in params_and_values]
```
(copied comment by @yizow from the previous PR)

`params, values = zip(*params_and_values)`

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -146,17 +269,68 @@ def make_error(error_code):
+
+
+
+def blocking_read(serial_conn):
```
(copied comment by @yizow from previous PR)

do we include the length of the message in the bytestream we receive?
How is this blocking? are you assuming serial_conn has a timeout on it?

This file doesnt even import serial. The above answers should be commented here, i feel.

nikitakit commented on hibike/hibike_process.py:
```diff
@@ -0,0 +1,186 @@
+                instruction_queues[uid_to_index[uid]].put(("read", args))            
+
+
+class DeviceWriteThread(threading.Thread):
```
(by @yizow, from previous PR)

Are we going to rewrite these classes to not extend `threading.Thread`?

nikitakit commented on hibike/hibike_process.py:
```diff
@@ -0,0 +1,186 @@
+
+    def run(self):
+        while True:
+            packet = hm.blocking_read(self.ser)
```
(by @yizow from previous PR)

again, i'm not too sure how this blocks.

karthik-shanmugam replied:
see hibike_message.py

nikitakit commented on hibike/hibike_process.py:
```diff
@@ -0,0 +1,186 @@
+
+if __name__ == "__main__":
+
+    # helper functions so we can spawn threads that try to
```
(by @yizow from previous PR)
threads that try to...?

nikitakit commented on hibike/list_devices.py:
```diff
@@ -0,0 +1,75 @@
+    for s in serials:
+        reading = hm.blocking_read(s)
+        if reading:
+            params, delay, device_type, year, id = struct.unpack("<HHHBQ", reading.getPayload())
```
(copied from previous PR)

This code looks like a duplicate of parse_subscription_response

nikitakit commented on hibike/spawn_virtual_devices.py:
```diff
@@ -0,0 +1,38 @@
+import csv
```
~~~Not reviewing this file now, pending resolving the issues raised in the last PR.~~~

File now reviewed.

nikitakit commented on hibike/hibike_message.py:
```diff
@@ -1,55 +1,57 @@
-  messageTypes["Error"] :                1
+# Dictionary of message types: message id
+messageTypes = {
+  "Ping" :                 0x10,
```
As I mentioned last PR, the move to python3 has given us access to an Enum class that can better be used for `messageTypes` and `errorCodes`. That said, I would be fine adding those to the issue tracker rather than addressing in this PR.

nikitakit commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,144 @@
+    newProcess = multiprocessing.Process(target=hibike_process, name="hibike_sim", args=[badThingsQueue, stateQueue, pipeFromChild])
+    newProcess.daemon = True
+    newProcess.start()
+    pipeToChild.send(["ready", []])
```
Again, what's the situation with the ready message, which is supposedly deprecated?

yizow replied:
this should be removed.

nikitakit commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,144 @@
+
+
+def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
+    while pipeFromChild.recv()[0] != "ready":
```
What is the status regarding the "ready" message, which is supposedly deprecated?

yizow replied:
This should just be removed.

nikitakit reviewed (changes requested):
Mostly LGTM, except that the virtual device comments from the previous PR have not been addressed yet.

In addition, I've copied over some comments from last PR that still haven't been addressed. Could you reply to them with a resolution?

You should also figure out what to do about misspelling "callibrate".

I also heard that the "ready" message was deprecated, but at this point I think the right resolution is to keep it in this PR, and then submit a future PR that makes sure that things are in sync with runtime.

ivanache committed:
Implemented Nikita's comments

nikitakit commented on hibike/spawn_virtual_devices.py:
```diff
@@ -0,0 +1,38 @@
+def get_virtual_ports():
+    socat = subprocess.Popen(shlex.split("socat -d -d pty,raw,echo=0 pty,raw,echo=0"), stderr=subprocess.PIPE)
+    popens.append(socat)
+    lines = [str(socat.stderr.readline()), str(socat.stderr.readline())]
```
(copied from previous PR)
`str` is not the right function for converting bytestrings to strings. Try `str(b"test")` in the REPL, it won't do the right thing.

If fixing bytestrings is indeed needed, you need to use `encode` to specify ascii, utf-8, or another encoding.

Given that `str` cannot possibly cast correctly, then either:
1. You can remove the cast
2. You should use the correct byte-to-str conversion

nikitakit replied:
`socat.stderr.readline()` returns a string, not bytes.

nikitakit commented on hibike/spawn_virtual_devices.py:
```diff
@@ -0,0 +1,38 @@
+def spawn_device(device_type):
+    p1, p2 = get_virtual_ports()
+    fname = os.path.join(os.path.dirname(__file__), "virtual_device.py")
+    device = subprocess.Popen(shlex.split("python3.5 %s -d %s -p %s" % (fname, device_type, p1)))
```
(copied from previous PR)
I think you should inserts some line breaks in here (maybe after some commas or parens). No need to use `\` or change the syntax, you can just use the already existing delimiters.

nikitakit commented on hibike/virtual_device.py:
```diff
@@ -0,0 +1,113 @@
+
+        # Send the written data, make sure you only send data for readable parameters
+        for index in range(len(write_params)):
+            while not hm.paramMap[device_id][write_tuples[index][0]][2]:
```
Does this loop even work correctly? It uses `write_tuples[index][0]` in the while loop condition, but then calls `del write_tuples[index]` in the body. I suspect you will end up entering the while loop once, deleting this object, and then crashing when you evaluate the loop condition a second time. Could you comment on this?

nikitakit replied:
This has now been addressed

nikitakit reviewed (changes requested):
Now also reviewed virtual devices (both `virtual_device.py` and `spawn_virtual_devices.py`).

All my comments on the previous version of `virtual_device.py` have now been addressed!

However I still have a few more changes I'd like to request (see the comments here)

ivanache committed:
Implemented the last of the suggested changes

Alramech commented on hibike/__init__.py:
```diff
@@ -0,0 +1,3 @@
+__all__ = ['hibike_simulator', 'hibike_message']
+from . import hibike_simulator
+from . import hibike_message
```
Newline

yizow replied:
newline at end of file.

Alramech commented on hibike/hibike.py:
```diff
@@ -134,6 +135,9 @@ def _updateContextFile(self, contextFile, newDeviceType):
 
 
+    def addVirtualPort(self, port):
+        self.virtuals.append((port, None, None))
```
What are the nones for? add comments

Alramech commented on hibike/hibike.py:
```diff
@@ -389,7 +393,7 @@ def handleInput(self, n):
         if not self.hibike.serialToUID:
             return
-        randIter = self.hibike.serialToUID.keys()
+        randIter = list(self.hibike.serialToUID.keys())
```
No longer an iter, rename to randlist

yizow commented on .gitignore:
```diff
@@ -390,3 +390,6 @@ $RECYCLE.BIN/
 .Trash-*
+
+# Hibike dynamic files
+hibike/virtual_devices.txt
```
shouldnt be necessary to modify the global gitignore file. 
Plus, you've already added this to hibike's own gitignore.

yizow commented on hibike/hibike_message.py:
```diff
@@ -112,31 +114,152 @@ def checksum(data):
+  """
+  params_and_values = sorted(params_and_values, key=lambda x: paramMap[device_id][x[0]][0])
+  params = [param[0] for param in params_and_values]
+  params_bitmask = encode_params(device_id, params)
```
this `params_bitmask` line seems to be repeated identically over and over again in all these functions. is there perhaps a way to deduplicate it, perhaps in whatever function is calling these functions?

karthik-shanmugam committed:
[HIBIKE] clean up markdown files

yizow commented on hibike/hibike_simulator.py:
```diff
@@ -1,133 +1,144 @@
+
+fake_uids = [0 << 72, 7 << 72]
+
+
```
so many extraneous new lines.

Alramech commented on hibike/hibike_message.py:
```diff
@@ -89,11 +91,11 @@ def getDeviceType(uid):
 # return bits[63: 0] of the UID
 def getID(uid):
-  return uid & 0xffffffffffffffff
+  return uid & 0xFFFFFFFFFFFFFFFF
```
Replace magic numbers with more appropriate constant

karthik-shanmugam replied:
It's a 64 bit bit-mask. Would replacing this with a variable actually make it more readable?

Alramech commented on hibike/hibike_message.py:
```diff
@@ -112,31 +114,152 @@ def checksum(data):
-  m_buff.append(chr(chk))
+  m_buff.append(chk)
   encoded = cobs_encode(m_buff)
   out_buf = bytearray([0x00, len(encoded)]) + encoded
```
magic number

karthik-shanmugam replied:
I could rename 0 to PACKET_DELIMITER_BYTE or something but I'm not sure if that makes this easier to read

nikitakit commented:
I see that a number of style issues remain, but functionality looks OK now. I would be fine merging the present PR, and then making the style changes in subsequent commits. I think we're at the point where we can accept this code, and help fix remaining style issues here while @karthik-shanmugam can refocus on the firmware/C code that's still not integrated.

karthik-shanmugam committed:
[HIBIKE] stop subclassing Thread and delete obsolete simulator

yizow commented on hibike/hibike_process.py:
```diff
@@ -60,61 +60,41 @@ def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
+        message_type = packet.getmessageID()
+        if message_type == hm.messageTypes["SubscriptionResponse"]:
+            params, delay, uid = hm.parse_subscription_response(packet)
+            uid = uid
```
no longer necessary.

yizow reviewed (changes requested):
Single line change requested. Otherwise LGTM

karthik-shanmugam committed:
[HIBIKE] remove redundant variable assignment

karthik-shanmugam committed:
[HIBIKE] minor cleanup

- remove obsolete hibike.py file
- add newline to end of __init__.py
- remove unused and undocumented "die" command in hibike_process
